### PR TITLE
Use more expressive notation for deriveSerializeWith

### PR DESCRIPTION
**Pull request by adithyaov**

### CODE REVIEW

File: core/src/Streamly/Internal/Data/Serialize/TH.hs [hunk 1]
```diff
@@ -381,44 +380,47 @@ mkSerializeDec (Config {..}) headTy cons = do
...
+  where
+
+    next (InstanceD mo preds headTy []) methods =
+        pure [InstanceD mo preds headTy methods]
+    next _ _ = errorUnsupported
+
+    getMainTypeName = go . unwrap
+
+    unwrap (AppT (ConT _) r) = r
+    unwrap _ = error "getMainTypeName.unwrap: Undefined instance declaration."
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/Serialize/TH.hs:418]: Should use the prefix `deriveSerializeWith` as that is what the user knows about. Also, does the message convey the specific error information that the user can understand, act upon and correct.

Comment (adithyaov) [reply core/src/Streamly/Internal/Data/Serialize/TH.hs:418]: Yes, we should be more specific.

File: core/src/Streamly/Internal/Data/Serialize/TH.hs [hunk 2]
```diff
@@ -412,14 +412,31 @@ deriveSerializeWith conf mDecs = do
...
-    go (ConT nm) = nm
-    go (AppT l _) = go l
-    go _ = error "getMainTypeName.go: Undefined instance declaration."
+    errorMessage dec =
+        error $ unlines
+            [ "Error: deriveSerializeWith:"
+            , ""
+            , ">> " ++ pprint dec
+            , ""
+            , "The above is not a proper instance declaration."
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/Serialize/TH.hs:421]: "proper" => valid

File: core/src/Streamly/Internal/Data/Serialize/TH.hs [hunk 3]
```diff
@@ -412,14 +412,31 @@ deriveSerializeWith conf mDecs = do
...
+            , "Any proper instance declaration without a body is valid."
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/Serialize/TH.hs:422]: proper => Haskell

**Review by harendra-kumar — APPROVED** [review]

**Pull request merged**